crates/apps/src/lib/node/ledger/shell/finalize_block.rs

```diff
@@ -652,6 +503,372 @@ where
+                    | Error::TxApply(protocol::Error::ReplayAttempt(_))
+            ) {
+                self.commit_batch_hash(tx_data.replay_protection_hashes);
+            } else if let Error::TxApply(protocol::Error::ReplayAttempt(_)) =
+                err
+            {
+                // Remove the wrapper hash but keep the inner tx
+                // hash. A replay of the wrapper is impossible since
+                // the inner tx hash is committed to storage and
+                // we validate the wrapper against that hash too
+                let header_hash = tx_data
+                    .replay_protection_hashes
+                    .expect("This cannot fail")
+                    .header_hash;
+                self.state
+                    .redundant_tx_hash(&header_hash)
+                    .expect("Error while marking tx hash as redundant");
+            }
+        }
+    }
+}
+
+struct TxData<'tx> {
+    is_atomic_batch: bool,
+    tx_header: &'tx namada::tx::Header,
+    commitments_len: u64,
+    tx_index: usize,
+    replay_protection_hashes: Option<ReplayProtectionHashes>,
+    consumed_gas: Gas,
+    height: BlockHeight,
+    wrapper_args: Option<WrapperArgs<'tx>>,
+}
+
+struct TxLogs<'finalize> {
+    tx_event: &'finalize mut Event,
+    stats: &'finalize mut InternalStats,
+    changed_keys: &'finalize mut BTreeSet<Key>,
+}
+
+// Temporary support type to update the tx logs. If the tx is confirmed this
+// gets merged to the non-temporary type
+struct TempTxLogs {
+    tx_event: Event,
+    stats: InternalStats,
+    changed_keys: BTreeSet<Key>,
+    other_events: Vec<Event>,
+}
+
+impl Default for TempTxLogs {
+    fn default() -> Self {
+        Self {
+            tx_event: Event::applied_tx(),
+            stats: Default::default(),
+            changed_keys: Default::default(),
+            other_events: Default::default(),
+        }
+    }
+}
+
+impl<'finalize> TempTxLogs {
+    // Consumes the temporary logs and merges them to confirmed ones. Pushes ibc
+    // and eth events to the finalize block response
+    fn commit(
+        self,
+        logs: &mut TxLogs<'finalize>,
+        response: &mut shim::response::FinalizeBlock,
+    ) {
+        logs.tx_event.attributes.extend(self.tx_event.attributes);
+        logs.stats.merge(self.stats);
+        logs.changed_keys.extend(self.changed_keys);
+        response.events.extend(self.other_events);
+    }
+
+    // Consumes the temporary logs and merges the statistics to confirmed ones.
+    // This is useful for failing atomic batches
+    fn commit_stats_only(self, logs: &mut TxLogs<'finalize>) {
+        logs.stats.merge(self.stats);
+    }
+
+    fn check_inner_results(
+        &mut self,
+        tx_result: &namada::tx::data::TxResult<protocol::Error>,
+        tx_header: &namada::tx::Header,
+        tx_index: usize,
+        height: BlockHeight,
+        commit_batch_hash: &mut bool,
+        is_any_tx_invalid: &mut bool,
```

Comment:
it's a small style issue, but passing these as mut refs instead of return value is not very readable because you don't know what the value is when it's being called without looking up the callers - you can just make some struct for these if you want to return them named

Reply:
I've tried to address this in 525d044c624cc4c189a845ed96eb5b7c584715b3